File: go/application/usecases/order_service.go
package usecases

import (
	"taller/application/ports/in"
	"taller/application/ports/out"
	"taller/domain/model"
	"taller/domain/service"
)

type OrderService struct {
	inventory       out.InventoryRepository
	orderRepository out.OrderRepository
	orderValidator  *service.OrderValidator
}

func NewOrderService(
	inventory out.InventoryRepository,
	orderRepository out.OrderRepository,
	orderValidator *service.OrderValidator,
) in.OrdersPort {
	return &OrderService{
		inventory:       inventory,
		orderRepository: orderRepository,
		orderValidator:  orderValidator,
	}
}

func (o *OrderService) GetOrders() ([]model.Order, error) {
	return o.orderRepository.GetOrders()
}

func (o *OrderService) ProcessOrder(order model.Order) error {
	coffeeBean, err := o.inventory.FindCoffeeBean(order.CoffeeBeanID)
	if err != nil {
		return err
	}

	if err = o.orderValidator.ValidateOrder(order, coffeeBean.AvailableGrams); err != nil {
		order.Status = model.Rejected
		return err
	}

	if err = o.inventory.RemoveStock(order.CoffeeBeanID, order.QuantityGrams); err != nil {
		return err
	}

	order.Status = model.Confirmed

	return o.orderRepository.Save(order)
}

File: typescript/src/application/usecases/order-service.ts
import { OrdersPort } from "../ports/in/orders-port";
import { OrderRepository } from "../ports/out/order-repository";
import { InventoryRepository } from "../ports/out/inventory-repository";
import { Order } from "../../domain/model/order";
import { OrderStatus } from "../../domain/model/order-status";
import { OrderValidator } from "../../domain/service/order-validator";

export class OrderService implements OrdersPort {
  private inventory: InventoryRepository;
  private orderRepository: OrderRepository;
  private orderValidator: OrderValidator;

  constructor(
    inventory: InventoryRepository,
    orderRepository: OrderRepository,
    orderValidator: OrderValidator
  ) {
    this.inventory = inventory;
    this.orderRepository = orderRepository;
    this.orderValidator = orderValidator;
  }

  getOrders(): Order[] {
    return this.orderRepository.getOrders();
  }

  processOrder(order: Order): void {
    const coffeeBean = this.inventory.findCoffeeBean(order.coffeeBeanId);

    this.orderValidator.validateOrder(order, coffeeBean.availableGrams);

    this.inventory.removeStock(order.coffeeBeanId, order.quantityGrams);

    order.status = OrderStatus.Confirmed;

    this.orderRepository.save(order);
  }
}
